File: bubble_sort.cpp
#include <iostream>
#include <vector>

using namespace std;

// Fungsi pembantu untuk mencetak array
void printArray(const vector<int>& arr) {
    cout << "[";
    for (size_t i = 0; i < arr.size(); i++) {
        cout << arr[i] << (i == arr.size() - 1 ? "" : ", ");
    }
    cout << "]" << endl;
}

void bubbleSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; i++) {
        cout << "Iterasi " << i + 1 << ":" << endl;
        bool swapped = false;
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
                swapped = true;
            }
        }
        printArray(arr);
        if (!swapped) break;
    }
}

int main() {
    vector<int> data = {78, 48, 10, 3, 36, 37, 43, 16, 34, 45, 25, 38, 72, 36, 53, 6};
    
    cout << "=== BUBBLE SORT ===" << endl;
    cout << "Jumlah Data: " << data.size() << endl;
    cout << "Data Awal: ";
    printArray(data);
    cout << "------------------------------" << endl;
    
    bubbleSort(data);
    
    cout << "------------------------------" << endl;
    cout << "Data Hasil Sorting: ";
    printArray(data);
    
    return 0;
}

File: bubble_sort.py
def bubble_sort(arr):
    n = len(arr)
    for i in range(n - 1):
        print(f"Iterasi {i + 1}:")
        swapped = False
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:
                # Tukar posisi
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        print(arr)
        # Jika tidak ada yang ditukar, berarti data sudah urut
        if not swapped:
            break
    return arr

# Data awal
data = [78, 48, 10, 3, 36, 37, 43, 16, 34, 45, 25, 38, 72, 36, 53, 6]

print("=== BUBBLE SORT ===")
print(f"Jumlah Data: {len(data)}")
print("Data Awal:", data)
print("-" * 30)

hasil = bubble_sort(data.copy())

print("-" * 30)
print("Data Hasil Sorting:", hasil)